### PR TITLE
Define cells to run as independent of selection

### PR DESCRIPTION
**Pull request by krassowski**
## References

Addresses #13361

Instead of relying on selection state for running cells (which is how individual run commands are implemented), save the cells to run before asynchronous actions to ensure that selection changes do not influence which cells get run.

## Code changes

- adds `Private.runSelected()` function
- adds public `NotebookActions.runCells()` function
- rewrites `notebook:restart-and-run-to-selected` and `notebook:restart-run-all` commands to use `NotebookActions.runCells`

## User-facing changes

When restarting and re-running up to selected cell, the cell which was selected during invoking the command (rather than the cell selected after restart) will be the one executed in the end.

## Backwards-incompatible changes

None I think?

### CODE REVIEW

File: packages/notebook/src/actions.tsx [hunk 1]
```diff
@@ -803,13 +838,9 @@ export namespace NotebookActions {
...
 
-    notebook.deselectAll();
-    for (let i = notebook.activeCellIndex; i < notebook.widgets.length; ++i) {
-      notebook.select(notebook.widgets[i]);
-    }
-
-    const promise = Private.runSelected(
+    const promise = Private.runCells(
       notebook,
+      notebook.widgets.slice(notebook.activeCellIndex, undefined),
```
Comment (fcollonval) [inline packages/notebook/src/actions.tsx:843]: Can be omitted

```suggestion
      notebook.widgets.slice(notebook.activeCellIndex),
```

File: packages/notebook/src/actions.tsx [hunk 2]
```diff
@@ -803,13 +838,9 @@ export namespace NotebookActions {
 
     const state = Private.getState(notebook);
 
-    notebook.deselectAll();
-    for (let i = notebook.activeCellIndex; i < notebook.widgets.length; ++i) {
-      notebook.select(notebook.widgets[i]);
-    }
-
-    const promise = Private.runSelected(
+    const promise = Private.runCells(
```
Comment (fcollonval) [inline packages/notebook/src/actions.tsx:841]: I have a change in behavior for this one. It does not activate the last cell after it is done (it keeps the selected cell activated).

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @krassowski 

The code looks good. However I spotted one regression. When calling run selected cell and all below, the last cell is not activated (the current select cell stays the one active)

File: packages/notebook/src/actions.tsx (hunk 2 above)
Comment (krassowski) [reply packages/notebook/src/actions.tsx:842]: Good catch, thank you! Addressed and tested in 202e958e967955d7a3d94a1d666dd7661a77a644.

**Review by fcollonval — APPROVED** [review]
Thanks @krassowski

**Pull request merged**